File: iscn_authenticator/engine.py
# iscn_authenticator/engine.py
"""Rule engine for applying validation rules to karyotype AST."""
from typing import Optional
from iscn_authenticator.models import KaryotypeAST, ValidationResult
from iscn_authenticator.rules.base import Rule


class RuleEngine:
    """Applies validation rules to a karyotype AST."""

    def __init__(self):
        self._rules: list[Rule] = []
        self._abnormality_rules: list[Rule] = []

    def add_rule(self, rule: Rule) -> None:
        """Add a rule to the engine."""
        self._rules.append(rule)

    def add_rules(self, rules: list[Rule]) -> None:
        """Add multiple rules to the engine."""
        self._rules.extend(rules)

    def add_abnormality_rule(self, rule: Rule) -> None:
        """Add an abnormality rule (applied to each abnormality)."""
        self._abnormality_rules.append(rule)

    def add_abnormality_rules(self, rules: list[Rule]) -> None:
        """Add multiple abnormality rules."""
        self._abnormality_rules.extend(rules)

    def validate(self, ast: KaryotypeAST) -> ValidationResult:
        """Validate a karyotype AST against all rules."""
        all_errors: list[str] = []

        # Apply AST-level rules
        for rule in self._rules:
            errors = rule.validate(ast, ast)
            all_errors.extend(errors)

        # Apply abnormality rules to each abnormality
        for abnormality in ast.abnormalities:
            for rule in self._abnormality_rules:
                errors = rule.validate(ast, abnormality)
                all_errors.extend(errors)

        return ValidationResult(
            valid=len(all_errors) == 0,
            errors=all_errors,
            parsed=ast
        )

File: packages/core/src/engine.ts
/**
 * Rule engine for applying validation rules to karyotype AST.
 * Port of iscn_authenticator/engine.py to TypeScript.
 */

import type { KaryotypeAST, ValidationResult, Rule } from "./types.ts";

export class RuleEngine {
  private rules: Rule[] = [];
  private abnormalityRules: Rule[] = [];

  addRule(rule: Rule): void {
    this.rules.push(rule);
  }

  addRules(rules: Rule[]): void {
    this.rules.push(...rules);
  }

  addAbnormalityRule(rule: Rule): void {
    this.abnormalityRules.push(rule);
  }

  addAbnormalityRules(rules: Rule[]): void {
    this.abnormalityRules.push(...rules);
  }

  validate(ast: KaryotypeAST): ValidationResult {
    const allErrors: string[] = [];

    // Apply AST-level rules
    for (const rule of this.rules) {
      const errors = rule.validate(ast, ast);
      allErrors.push(...errors);
    }

    // Apply abnormality rules to each abnormality
    for (const abnormality of ast.abnormalities) {
      for (const rule of this.abnormalityRules) {
        const errors = rule.validate(ast, abnormality);
        allErrors.push(...errors);
      }
    }

    return {
      valid: allErrors.length === 0,
      errors: allErrors,
      parsed: ast,
    };
  }
}
